scripts/extract_implications.lean

```diff
@@ -1,20 +1,47 @@
-import Lean.Meta.Basic
-import Lean.Util
+import Batteries.Data.Array.Basic
+import Batteries.Tactic.Lint.Frontend
+import Cli.Basic
+import Lean.Util.SearchPath
+import equational_theories.Closure
 
-import equational_theories.EquationalResult
-
-open Lean
-
-unsafe def main (args : List String) : IO Unit := do
-  let module := ← match args with
-    | [] => pure `equational_theories
-    | [mod] => pure mod.toName
-    | _ => throw <| IO.userError "usage: extract_implications MODULE"
+open Lean Core Elab Cli
 
+def generateOutput (inp : Cli.Parsed) : IO UInt32 := do
+  let some modules := inp.variableArgsAs? ModuleName |
+    inp.printHelp
+    return 1
+  if modules.isEmpty then
+    inp.printHelp
+    return 2
+  let include_conj := inp.hasFlag "conjecture"
+  let include_impl := inp.hasFlag "closure"
   searchPathRef.set compile_time_search_path%
 
-  withImportModules #[{module}] {} (trustLevel := 1024) fun env =>
+  unsafe withImportModules (modules.map ({module := ·})) {} (trustLevel := 1024) fun env =>
     let ctx := {fileName := "", fileMap := default}
     let state := {env}
-    Prod.fst <$> (Meta.MetaM.toIO · ctx state)  do
-      println! (Lean.toJson (← Result.collectResults)).pretty
+    Prod.fst <$> (Meta.MetaM.toIO · ctx state) do
+      let mut rs ← Result.extractEquationalResults
+      if !include_conj then
+        rs := rs.filter (·.proven)
+      let rs' := rs.map (·.variant)
+      let mut rs' := if include_impl then Closure.closure rs' else Closure.toEdges rs'
+      for edge in rs' do
+        if edge.isTrue then IO.println s!"{edge.lhs} → {edge.rhs}"
```

Comment:
This is a reasonable format to print in, but it's sad to lose the JSON format, which other tools (especially those not written in Lean) might want.

Maybe we should split the functionality into multiple executables? Or maybe there should be a CLI flag specifying the output format?

Reply:
I added the ability to output to JSON back, but since it takes a lot more time I added a CLI flag for it.